x/auth/vesting/types/vesting_account.go

```diff
@@ -611,3 +589,412 @@ func marshalYaml(i interface{}) (interface{}, error) {
+	capPeriods := []Period{
+		{
+			Length: 0,
+			Amount: totalVested,
+		},
+	}
+	_, _, newLockupPeriods := ConjunctPeriods(tva.StartTime, tva.StartTime, tva.LockupPeriods, capPeriods)
+
+	_, _, newCombinedPeriods := ConjunctPeriods(tva.StartTime, tva.StartTime, newLockupPeriods, newVestingPeriods)
+
+	// Now construct the new account state
+	tva.OriginalVesting = totalVested
+	tva.EndTime = t
+	tva.LockupPeriods = newLockupPeriods
+	tva.VestingPeriods = newVestingPeriods
+	tva.CombinedPeriods = newCombinedPeriods
+	// DelegatedVesting will be adjusted elsewhere
+
+	return tva, totalUnvested
+}
+
+// Clawback transfers unvested tokens in a TrueVestingAccount to dest.
+// Future vesting events are removed. Unstaked tokens are simply sent.
+// Unbonding and staked tokens are transferred with their staking state
+// intact.  Account state is updated to reflect the removals.
+func (tva TrueVestingAccount) Clawback(ctx sdk.Context, dest sdk.AccAddress, ak AccountKeeper, bk BankKeeper, sk StakingKeeper) error {
+	updatedAcc, toClawBack := tva.ComputeClawback(ctx.BlockTime().Unix())
+	if toClawBack.IsZero() {
+		return nil
+	}
+	addr := updatedAcc.GetAddress()
+
+	// Write now now so that the bank module sees unvested tokens are unlocked.
+	// Note that all store writes are aborted if there is a panic, so there is
+	// no danger in writing incomplete results.
+	ak.SetAccount(ctx, &updatedAcc)
+
+	// Now that future vesting events (and associated lockup) are removed,
+	// the balance of the account is unlocked and can be freely transferred.
+	spendable := bk.SpendableCoins(ctx, addr)
+	toXfer := coinsMin(toClawBack, spendable)
+	err := bk.SendCoins(ctx, addr, dest, toXfer)
+	if err != nil {
+		return err // shouldn't happen, given spendable check
+	}
+	toClawBack = toClawBack.Sub(toXfer)
+
+	// We need to traverse the staking data structures to update the
+	// vesting account bookkeeping, and to recover more funds if necessary.
+	// Staking is the only way unvested tokens should be missing from the bank balance.
+	bondDenom := sk.BondDenom(ctx)
+
+	// If we need more, transfer UnbondingDelegations.
+	want := toClawBack.AmountOf(bondDenom)
+	unbondings := sk.GetUnbondingDelegations(ctx, addr, math.MaxUint16)
+	for _, unbonding := range unbondings {
+		transferred := sk.TransferUnbonding(ctx, addr, dest, sdk.ValAddress(unbonding.ValidatorAddress), want)
+		want = want.Sub(transferred)
+		if !want.IsPositive() {
+			break
+		}
+	}
+
+	// If we need more, transfer Delegations.
+	if want.IsPositive() {
+		delegations := sk.GetDelegatorDelegations(ctx, addr, math.MaxUint16)
+		for _, delegation := range delegations {
+			validatorAddr, err := sdk.ValAddressFromBech32(delegation.ValidatorAddress)
+			if err != nil {
+				panic(err) // shouldn't happen
+			}
+			validator, found := sk.GetValidator(ctx, validatorAddr)
+			if !found {
+				panic("validator not found") // shouldn't happen
+			}
+			wantShares, err := validator.SharesFromTokensTruncated(want)
```

Comment:
This should handle a slashing event correctly

Reply:
Do you mean "this should be changed so that it handles slashing" or "it's great that this already handles slashing"?  It looks like validators are never removed if any delegator is still holding shares, but I made the code tolerate a missing validator, reading the delegation as having no value, both here and in similar following code..